### PR TITLE
service/frd: return cfg username on GetMyScreenName

### PR DESCRIPTION
**Pull request by vitor-k**
Some games apparently use this for the multiplayer display name.
According to [this post](https://community.citra-emu.org/t/name-specified-in-system-settings-doesnt-change-multiplayer-name/414197) on the forums, some examples of games that do that are Mario Sports Superstars and Metroid Prime Federation Force.
I don't have any affected games so I haven't tested this.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/5792)

### CODE REVIEW

**Comment by vitor-k** [comment]
From the testing from the forum user, this change does indeed result in the name defined in the cfg being used for the multiplayer display name.

**Review by B3n30 — APPROVED** [review]

File: src/core/hle/service/frd/frd.cpp [hunk 1]
```diff
@@ -93,16 +95,21 @@ void Module::Interface::GetMyScreenName(Kernel::HLERequestContext& ctx) {
     IPC::RequestBuilder rb = rp.MakeBuilder(7, 0);
 
     struct ScreenName {
-        std::array<char16_t, 12> name;
+        // 20 bytes according to 3dbrew
+        char16_t name[10];
```
Comment (zhaowenlan1779) [inline src/core/hle/service/frd/frd.cpp:99]: Should still use a `std::array` here

File: src/core/hle/service/frd/frd.cpp [hunk 2]
```diff
@@ -93,16 +95,21 @@ void Module::Interface::GetMyScreenName(Kernel::HLERequestContext& ctx) {
...
     };
 
-    // TODO: (mailwl) get the name from config
-    ScreenName screen_name{u"Citra"};
+    auto cfg = Service::CFG::GetModule(Core::System::GetInstance());
+    ASSERT_MSG(cfg, "CFG Module missing!");
+    auto username = cfg->GetUsername();
+    ASSERT_MSG(username.length() <= 10, "Username longer than expected!");
+    ScreenName screen_name{0};
```
Comment (zhaowenlan1779) [inline src/core/hle/service/frd/frd.cpp:106]:
```suggestion
    ScreenName screen_name{};
```
for consistency

File: src/core/hle/service/frd/frd.cpp [hunk 3]
```diff
@@ -93,16 +95,21 @@ void Module::Interface::GetMyScreenName(Kernel::HLERequestContext& ctx) {
...
 
     struct ScreenName {
-        std::array<char16_t, 12> name;
+        // 20 bytes according to 3dbrew
+        char16_t name[10];
     };
 
-    // TODO: (mailwl) get the name from config
-    ScreenName screen_name{u"Citra"};
+    auto cfg = Service::CFG::GetModule(Core::System::GetInstance());
```
Comment (zhaowenlan1779) [inline src/core/hle/service/frd/frd.cpp:102]: I don't really like how this is using the global System. You should store a System reference in `Service::FRD::Module` (pass it in from `InstallInterfaces`) and use it instead.

**Review by BreadFish64 — APPROVED** [review]

File: src/core/hle/service/frd/frd.cpp [hunk 4]
```diff
@@ -93,16 +97,21 @@ void Module::Interface::GetMyScreenName(Kernel::HLERequestContext& ctx) {
...
+        // 20 bytes according to 3dbrew
+        std::array<char16_t, 10> name;
     };
 
-    // TODO: (mailwl) get the name from config
-    ScreenName screen_name{u"Citra"};
+    auto cfg = Service::CFG::GetModule(frd->system);
+    ASSERT_MSG(cfg, "CFG Module missing!");
+    auto username = cfg->GetUsername();
+    ASSERT_MSG(username.length() <= 10, "Username longer than expected!");
```
Comment (zhaowenlan1779) [inline src/core/hle/service/frd/frd.cpp:107]: 3dbrew says that the username field in config savegame is 0x1C bytes long (which is 14 utf-16 characters). Although Citra's configuration dialog (and the 3DS's, I suppose?) limit it to 10 characters, I think this shouldn't be an assertion here. I would say just report a error and truncate the string.

Comment (vitor-k) [reply src/core/hle/service/frd/frd.cpp:107]: According to [this](https://www.3dbrew.org/wiki/Config_Savegame#0x000A0000_Block), username goes from 0x0 to 0x13 of the block, with the rest being zeros and some name validation stuff.
The UsernameBlock struct we use also has only 20 bytes to the username, 10 utf-16 characters.

**Review by bunnei — APPROVED** [review]

**Pull request merged**